playground/serve.go

```diff
@@ -16,27 +16,237 @@
+	Components []byte
+}
+
 func InitPlayground() {
 	http.HandleFunc("/playground/fetch", handler)
+	http.HandleFunc("/playground/amp-component-versions", components)
+	http.HandleFunc("/playground/amp-component-versions-task", componentsTask)
 	validRequestUrlOrigins = map[string]bool{
 		"ampbyexample.com":                     true,
 		"ampstart.com":                         true,
 		"ampstart-staging.firebaseapp.com":     true,
 		"localhost:8080":                       true,
 		"amp-by-example-staging.appspot.com":   true,
 		"amp-by-example-sebastian.appspot.com": true,
+		"0.1.0.1": true,
+	}
+}
+
+func InitializeComponents(r *http.Request) {
+	getComponentsAndUpdateIfStale(r)
+}
+
+func components(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		http.Error(w, "only GET request supported", http.StatusBadRequest)
+		return
+	}
+	if r.Header.Get("x-requested-by") != "playground" {
+		http.Error(w, "x-requested-by invalid", http.StatusBadRequest)
+		return
+	}
+	components := getComponentsAndUpdateIfStale(r)
+	if components == nil {
+		http.Error(w, "Server error", http.StatusServiceUnavailable)
+		return
+	}
+
+	w.Header().Set("Content-type", "application/json")
+	w.Write(components.Components)
+}
+
+func getComponentsAndUpdateIfStale(r *http.Request) *AmpComponentsList {
+	curTime := int(time.Now().Unix())
+	ctx := appengine.NewContext(r)
+	latest, _ := getComponentsFromMemCache(ctx)
+	// latest could be nil if not in memcache or datastore.
+	var timestamp int
+	if latest != nil {
+		timestamp = latest.Timestamp
+	}
+
+	if curTime-timestamp > COMPONENTS_UPDATE_FREQ_SECONDS {
+		createTaskQueueUpdate(ctx, timestamp)
+	}
+	return latest
+}
+
+func createTaskQueueUpdate(ctx context.Context, timestamp int) {
+	t := taskqueue.NewPOSTTask("/playground/amp-component-versions-task",
+		url.Values{})
+	// Setting the name explicitly means that only one task will ever
+	// be in the queue, even if attempted by separate instances.
+	// The name chosen includes the timestamp of the last known
+	// components list (or 0 for the first attempt), which means
+	// that only one update task can be created for each version
+	// known.
+	t.Name = fmt.Sprintf("amp-components-list-last-known-%d", timestamp)
+	minBackoff, _ := time.ParseDuration("10s")
+	t.RetryOptions = &taskqueue.RetryOptions{
+		RetryLimit: 5,
+		MinBackoff: minBackoff,
+	}
+	log.Infof(ctx, "Adding to taskqueue: %s", t.Name)
+	taskqueue.Add(ctx, t, "")
+}
+
+func componentsTask(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+	_, err := updateComponents(ctx)
+	if err != nil {
+		http.Error(w, "Server error", http.StatusInternalServerError)
+		return
+	}
+	w.Write([]byte("Updated components"))
+}
+
+func updateComponents(ctx context.Context) (bool, error) {
+	log.Infof(ctx, "Fetching components list from GitHub")
+	list, err := fetchComponents(ctx)
+	if err != nil {
+		return false, err
+	}
+	addComponentsToStores(ctx, list)
+	return true, nil
```

Comment:
Remove the bool return value as it's not needed.

Reply:
Done